[WIP] Add mangle/demangle helper functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122796
* #123129
* #123110
* #123109
* #123108

- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

zou3519 committed:
Add unique_name/mangle/demangle helper functions

- Given a Python lambda or function (which may be nested), unique_name
  produces a name that uniquely identifies it.
- `mangle` mangles any string that matches `[A-z0-9.<>@]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation

In the next PR up in the stack, we use `mangle(unique_name(fn))` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add unique_name/mangle/demangle helper functions"


- Given a Python lambda or function (which may be nested), unique_name
  produces a name that uniquely identifies it.
- `mangle` mangles any string that matches `[A-z0-9.<>@]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation

In the next PR up in the stack, we use `mangle(unique_name(fn))` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add unique_name/mangle/demangle helper functions"


- Given a Python lambda or function (which may be nested), unique_name
  produces a name that uniquely identifies it.
- `mangle` mangles any string that matches `[A-z0-9.<>@]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation

In the next PR up in the stack, we use `mangle(unique_name(fn))` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/utils.py:
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
+    - a function is identified by its __module__ and __qualname__
+    - lambda requires the object's id to distinguish
+    """
+    is_lambda = fn.__name__ == "<lambda>"
```
nit: check that the input is actually a callable?
Also is a Callable guaranteed to have a `__name__`/`__module__`/`__qualname__` attributes?

albanD commented on torch/_library/utils.py:
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
+def mangle(value: str) -> str:
+    """Mangles a string into a usable Python identifier.
+
+    https://docs.python.org/3/reference/lexical_analysis.html#identifiers
```
I'm surprised there is no generic tool to make any string into a indentifier...

Skylion007 replied:
This doesn't handle if a value is a keyword, right? That can be detect with lib/keyword.py

Skylion007 replied:
There is does github gist that seems to cover some more edgecases https://gist.github.com/JamesPHoughton/3a3f87c6662bf5c9eccc9f2206e228fd

albanD commented on torch/_library/utils.py:
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
+    return value
+
+
+def unique_name(fn: Callable) -> str:
```
callable_unique_name ?

albanD commented on test/test_custom_ops.py:
```diff
@@ -721,6 +722,40 @@ def _generate_examples(self, typ):
+            demangled = utils.demangle(mangled)
+            self.assertEqual(demangled, value)
+
+    def test_unique_name(self):
```
I would also test this with:
- method
- classmethod
- staticmethod
- builtin
- builtin method

ezyang commented:
@zou3519 and I talked about this in our 1:1, and the conclusion was that we only really need mangling for top level identifiers, and a special case for one shot inner functions just for ease of testing.

Skylion007 commented on torch/_library/utils.py:
```diff
@@ -156,3 +157,67 @@ def zip_schema(schema, args, kwargs):
+    - ZQ.ZQ -> ZZQXZZQ
+    - <lambda> -> Z0lambdaZ1
+    """
+    assert re.match(r"^[A-z0-9.<>@_]+$", value), "invalid value"
```
Did you intend to include ```ASCII values from 91 to 96 ([, \, ], ^, _, and `)``` Or should this be replaced too?

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
I am planning on turning this PR into the "automatic namespace support for custom_op" PR. No need to review yet

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[WIP] Add mangle/demangle helper functions"


- `mangle` mangles any string that matches `[A-z0-9.<>]+` into a valid
  Python identifier (e.g. has no special chars aside from underscore).
  We optimize for readability of the string in common cases. `demangle`
  is the inverse operation
- Why do we need <>? It's needed for nested functions.

In a future PR up in the stack, we use `mangle` to
generate a name that can be used an operator name. I split off the
mangling because that PR was getting long.

Test Plan:
- new tests

[ghstack-poisoned]
